core/lib/state/src/rocksdb/mod.rs

```diff
@@ -131,23 +199,162 @@ impl RocksdbStorage {
+                    StateKeeperColumnFamily::State,
+                    Self::ENUM_INDEX_MIGRATION_CURSOR,
+                    next_key.as_bytes(),
+                );
+            }
+            _ => {
+                write_batch.put_cf(
+                    StateKeeperColumnFamily::State,
+                    Self::ENUM_INDEX_MIGRATION_FINISHED,
+                    &[1],
+                );
+                write_batch.delete_cf(
+                    StateKeeperColumnFamily::State,
+                    Self::ENUM_INDEX_MIGRATION_CURSOR,
+                );
+                tracing::info!("RocksDB enum index migration finished");
+            }
+        }
+        self.db
+            .write(write_batch)
+            .expect("failed to save state data into rocksdb");
+        tracing::info!(
+            "RocksDB enum index migration chunk took {:?}, migrated {} keys",
+            started_at.elapsed(),
+            keys.len()
+        );
     }
 
     fn read_value_inner(&self, key: &StorageKey) -> Option<StorageValue> {
         let cf = StateKeeperColumnFamily::State;
         self.db
             .get_cf(cf, &Self::serialize_state_key(key))
             .expect("failed to read rocksdb state value")
-            .map(|value| H256::from_slice(&value))
+            .map(|value| H256::from_slice(&value[..32]))
     }
 
-    /// Processes storage `logs` produced by transactions.
-    fn process_transaction_logs(&mut self, updates: &HashMap<StorageKey, H256>) {
-        for (&key, &value) in updates {
-            if !value.is_zero() || self.read_value_inner(&key).is_some() {
-                self.pending_patch.state.insert(key, value);
-            }
-        }
+    fn read_state_value(&self, key: &StorageKey) -> Option<StateValue> {
+        let cf = StateKeeperColumnFamily::State;
+        self.db
+            .get_cf(cf, &Self::serialize_state_key(key))
+            .expect("failed to read rocksdb state value")
+            .map(|value| StateValue::deserialize(&value))
+    }
+
+    /// Returns storage logs to apply.
+    fn process_transaction_logs(
+        &self,
+        updates: HashMap<StorageKey, H256>,
+    ) -> HashMap<StorageKey, StateValue> {
+        updates
+            .into_iter()
+            .filter_map(|(key, new_value)| {
+                if let Some(state_value) = self.read_state_value(&key) {
+                    Some((
+                        key,
+                        StateValue {
+                            value: new_value,
+                            enum_index: state_value.enum_index,
+                        },
+                    ))
+                } else {
+                    (!new_value.is_zero()).then_some((
+                        key,
+                        StateValue {
+                            value: new_value,
+                            enum_index: None,
+                        },
+                    ))
+                }
+            })
+            .collect()
```

Comment:
Nit: AFAICT, you don't need to collect the iterator here; instead, you could return `impl Iterator` directly (it would have `&self` lifetime, but this looks OK).